internal/state/snapshot.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package state
+
+import (
+	"github.com/siderolabs/gen/maps"
+	"github.com/siderolabs/gen/xslices"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	storagepb "github.com/siderolabs/discovery-service/api/storage"
+)
+
+// Snapshot takes a snapshot of the current state.
+//
+// Implements storage.Snapshotter interface.
+func (state *State) Snapshot() *storagepb.StateSnapshot {
+	var clusterSnapshots []*storagepb.ClusterSnapshot
+
+	state.clusters.Range(func(_ string, cluster *Cluster) bool {
+		clusterSnapshots = append(clusterSnapshots, snapshotCluster(cluster))
+
+		return true
+	})
+
+	return &storagepb.StateSnapshot{
```

Comment:
can we do a benchmark of how many allocations happen on this path -- from the state snapshot to the final on-disk representation. 

I have some fears about protobuf

Reply:
plus it almost feels like we could do better job if we write cluster-by-cluster without trying to put all clusters at the same time to memory.

so there's no `StateSnapshot`, but a list of protobufs for each `Cluster`